adafruit_pm25/uart.py

```diff
@@ -46,38 +51,215 @@
+
+        Device isn't fully available until 3 seconds after wake is received
+
+        This command does not trigger a response, so we don't need to pull one off the buffer
+        """
+        self._uart.reset_input_buffer()
+        self._uart.write(self._build_cmd_frame(PLANTOWER_CMD_WAKEUP))
+        time.sleep(3)
+
+    def cmd_passive_read(self):
+        """
+        Sends command to request a data frame whlie in 'passive' mode and immediately reads in frame
+        """
+        self._uart.reset_input_buffer()
+        self._uart.write(self._build_cmd_frame(PLANTOWER_CMD_READ))
+
+    def pin_reset(self):
+        """
+        Resets device via RESET pin, but only if pin has been assigned
+
+        Reset via pin requires about 3 seconds before the device becomes available
+        """
+        if self._reset_pin is not None:
+            self._reset_pin.value = False
+            time.sleep(0.01)
+            self._reset_pin.value = True
+            time.sleep(3)
+            if self._mode == "passive":
+                self.cmd_mode_passive()
+
+    def pin_sleep(self):
+        """
+        Sleeps device via SET pin, but only if pin has been assigned
+        """
+        if self._set_pin is not None and self._set_pin.value:
+            self._set_pin.value = False
+
+    def pin_wakeup(self):
+        """
+        Wakes device via SET pin, but only if pin has been assigned
+
+        Wakeup from sleep via pin takes about 3 seconds before device is available
+        """
+        if self._set_pin is not None and not self._set_pin.value:
+            self._set_pin.value = True
+            time.sleep(3)
+
+    @staticmethod
+    def _build_cmd_frame(cmd_bytes):
+        """
+        Builds a valid command frame byte array with checksum for given command bytes
+        """
+        if len(cmd_bytes) != 3:
+            raise RuntimeError("Malformed command frame")
+        cmd_frame = bytearray()
+        cmd_frame.extend(PLANTOWER_HEADER)
+        cmd_frame.extend(cmd_bytes)
+        cmd_frame.extend(sum(cmd_frame).to_bytes(2, "big"))
+        return cmd_frame
+
+    def _read_uart(self):
+        # Disable too-many violations for this method, since it actually needs these branches
+        # pylint: disable=too-many-nested-blocks,too-many-branches
+        """
+        Reads a single frame via UART
+
+        Ignores bytes that are not frame headers to avoid reading in frames mid-stream
+        """
+        uart_timeouts = 0
+        error_count = 0
+        first_bytes_tried = 0
         while True:
-            b = self._uart.read(1)
-            if not b:
-                raise RuntimeError("Unable to read from PM2.5 (no start of frame)")
-            if b[0] == 0x42:
-                break
-        self._buffer[0] = b[0]  # first byte and start of frame
+            serial_data = bytearray()
+            first_byte = self._uart.read(1)
+            if first_byte is not None:
```

Comment:
I just tested a timeout at REPL. I've unplugged PMS5003, set timeout to 4 on UART and read() gives me `b''` whereas this code is expecteding `None`. `RX` line would be floating here, I think.

Reply:
Another good catch. I still don't have my bench set up again unfortunately.

Do you have a suggestion for a better way to handle the initial "did we get any actaul data" check?

I wasn't aware that the `uart` driver could send us an empty byte array in these scenarios. Maybe we could just check for both `is not None` (to cover normal, non-floating RX scenarios) as well as the content (or lack thereof) of the byte array we get back from `uart.read(1)` (for device-disconnect/floating RX scenarios)?

I feel like this should be a problem for many Circuitpython drivers like this that need to do incremental per-byte reads to find frame starts. Or is this the only one? :smile: